src/Build/Instance/ProjectItemInstance.cs

```diff
@@ -1076,6 +1086,28 @@ internal int DirectMetadataCount
                 }
             }
 
+            /// <summary>
+            /// Used to return metadata from another AppDomain. Can't use yield return because the
+            /// generated state machine is not marked as [Serializable], so we need to allocate.
+            /// </summary>
+            /// <param name="list">The source list to return metadata from.</param>
+            /// <returns>An array of string key-value pairs representing metadata.</returns>
+            private IEnumerable<KeyValuePair<string, string>> EnumerateMetadataEager(CopyOnWritePropertyDictionary<ProjectMetadataInstance> list)
+            {
+                var result = new List<KeyValuePair<string, string>>(list.Count);
+
+                foreach (var projectMetadataInstance in list)
+                {
+                    if (projectMetadataInstance != null)
+                    {
+                        result.Add(new KeyValuePair<string, string>(projectMetadataInstance.Name, projectMetadataInstance.EvaluatedValue));
+                    }
+                }
+
+                // Probably better to send the raw array across the wire even if it's another allocation.
+                return result.ToArray();
```

Comment:
Alternative: allocate array of length list.Count then send it.
Advantage: only one allocation
Possible disadvantage: If there are null instances, there would be some nulls at the end
Workaround with reduced speed: count nulls before allocating the array and allocate an appropriately-sized array.

Reply:
Yes, exactly. Those potential null instances break that. In the two other cases I directly allocate the array, but here I have to use a list.

Good news is that this is only where AppDomains are involved, and we don't really care about perf in these isolated scenarios. The mainline scenarios are not affected.